File: ts/packages/protocol/src/transport.ts
import type { Envelope } from './envelope.js';

/** Adapter-side dispatch fn: returns Response envelope for requests, null for notifications. */
export type Dispatcher = (env: Envelope) => Promise<Record<string, unknown> | null>;

export interface SendOptions {
  timeoutMs: number;
}

export interface Transport {
  /** Register a handler for inbound traffic on a given service name. */
  registerService(name: string, dispatch: Dispatcher): Promise<void>;
  /** Send a request envelope; resolve with the response envelope. */
  send(env: Record<string, unknown>, opts: SendOptions): Promise<Record<string, unknown>>;
  /** Send a fire-and-forget notification envelope. */
  notify(env: Record<string, unknown>): Promise<void>;
  /** Spin up loops; idempotent. */
  start(): Promise<void>;
  /** Drain in-flight + close; idempotent. */
  stop(): Promise<void>;
}

File: py/packages/protocol/src/clamator_protocol/transport.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any, Protocol, runtime_checkable

from .envelope import Envelope

Dispatcher = Callable[[Envelope], Awaitable[dict[str, Any] | None]]


@runtime_checkable
class Transport(Protocol):
    async def register_service(self, name: str, dispatch: Dispatcher) -> None: ...
    async def send(self, env: dict[str, Any], *, timeout: float) -> dict[str, Any]: ...
    async def notify(self, env: dict[str, Any]) -> None: ...
    async def start(self) -> None: ...
    async def stop(self) -> None: ...
